File: n queen/n_queen_backtracking.js
/**
 * @param {number} n
 * @return {string[][]}
 */

var isValid = function (board, row, col) {
  for (var i = 0; i < board.length; i++) {
    if (board[i][col] === 'Q') {
      return false;
    }
  }
  for (var i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
    if (board[i][j] === 'Q') {
      return false;
    }
  }
  for (var i = row - 1, j = col + 1; i >= 0 && j < board.length; i--, j++) {
    if (board[i][j] === 'Q') {
      return false;
    }
  }
  return true;
};

var backtracking = function (board, row, result) {
  if (row === board.length) {
    var temp = [];
    for (var i = 0; i < board.length; i++) {
      temp.push(board[i].join(''));
    }
    result.push(temp);
    return;
  }
  for (var i = 0; i < board.length; i++) {
    if (isValid(board, row, i)) {
      board[row][i] = 'Q';
      backtracking(board, row + 1, result);
      board[row][i] = '.';
    }
  }
};

var solveNQueens = function (n) {
  var result = [];
  var board = [];
  for (var i = 0; i < n; i++) {
    board[i] = [];
    for (var j = 0; j < n; j++) {
      board[i][j] = '.';
    }
  }
  backtracking(board, 0, result);

  return result;
};

File: n queen/n_queen_backtracking.py
from typing import List

def solveNQueens(self, n: int) -> List[List[str]]:
      board = [['.' for _ in range(n)] for _ in range(n)]
      column = set()
      positive_diagnol = set()
      negative_diagnol = set()
      result = []

      def backtrack(row):
        if row == n:
          result.append([''.join(row) for row in board])
          return 

        for col in range(n):
          if col in column or row + col in positive_diagnol or row - col in negative_diagnol:
            continue

          board[row][col] = 'Q'
          column.add(col)
          positive_diagnol.add(row + col)
          negative_diagnol.add(row - col)

          backtrack(row + 1)

          board[row][col] = '.'
          column.remove(col)
          positive_diagnol.remove(row + col)
          negative_diagnol.remove(row - col)
        
      backtrack(0)
      return result
